src/Concretize.hs

```diff
@@ -36,347 +59,398 @@ data Level = Toplevel | Inside
 -- |  Both of these results are returned in a tuple: (<new xobj>, <dependencies>)
 concretizeXObj :: Bool -> TypeEnv -> Env -> [SymPath] -> XObj -> Either TypeError (XObj, [XObj])
 concretizeXObj allowAmbiguityRoot typeEnv rootEnv visitedDefinitions root =
-  case runState (visit allowAmbiguityRoot Toplevel rootEnv root) [] of
+  case runState (visit (visitedDefinitions ++ [getPath root]) allowAmbiguityRoot Toplevel typeEnv rootEnv root) [] of
     (Left err, _) -> Left err
     (Right xobj, deps) -> Right (xobj, deps)
+
+--------------------------------------------------------------------------------
+-- Visit functions
+--
+-- These functions take a state and supporting information and gradually
+-- convert generically typed xobjs into concretely typed xobjs.
+--
+-- The functions prefixed "visit" primarily just recur into the component parts
+-- of different Carp forms while the "concretizeType..." functions perform the
+-- actual type conversion work.
+
+-- | The type of visit functions. These functions convert the types of the
+-- components of a form into concrete types and take the following arguments:
+--   - A bool indicating whether or not type variables are allowed
```

Comment:
The first argument ([SymPath]) isn't covered in this description.

Reply:
good catch! fixed